simpledotorg/simple-server
# Chartjs: Create baseConfig - line charts and combo bar&line charts baseConfig implementation [DM reports only]

**Story card:** [sc-9576](https://app.shortcut.com/simpledotorg/story/9576/phase-1-dry-chartjs-config-dm-only-line-graph-all-charts-with-axes)

## Because

Large amounts of chartjs config duplication

## This addresses

This is part of a multiphase update to the chartjs 'reports.js' file.

### **Overall aim**: 
- to reduce code duplication and update to v4.0.1 at end of epic.

### **Aim of this PR**: 
1. Create a base config file for all line charts and combo line&bar charts on the dashboard
2. Base config will follow the styling currently used for said charts
3. Implement base config fully (or partially) on said charts shown on the 'Diabetes' reports page (only diabetes for now)

Charts of focus in this PR: 
These are screenshots from before this PR (master)
<img width="696" alt="Screenshot 2022-11-30 at 18 52 43" src="https://user-images.githubusercontent.com/9541902/204807563-eeec17a2-ced0-4617-8ce0-6a5bf9b75726.png">
<img width="693" alt="Screenshot 2022-11-30 at 18 52 49" src="https://user-images.githubusercontent.com/9541902/204807572-7acc3817-12e2-478c-b83d-6d4ef6c8fbaa.png">
<img width="699" alt="Screenshot 2022-11-30 at 18 52 54" src="https://user-images.githubusercontent.com/9541902/204807585-35072f2a-77f7-4d26-a6e4-15be15f0cc63.png">
<img width="697" alt="Screenshot 2022-11-30 at 18 53 14" src="https://user-images.githubusercontent.com/9541902/204807591-e8f30e70-2369-4a6e-bd04-5176c0ae27f5.png">
<img width="700" alt="Screenshot 2022-11-30 at 18 53 29" src="https://user-images.githubusercontent.com/9541902/204807602-096341d6-22ed-4d71-a644-bc77bdcac100.png">

### Why partial implementation on some chart?
There is a difficulty in fully implementing the base config on charts showing more than a one dataset:
- Registrations and follow-ups
- Blood sugar 200-299 or ≥300

This is due to the structure of the 'scales' object [its an array] and the way withMerge works [specifically around merging arrays].
withMerge combines arrays making it hard to overwrite a specific key value.
In this case the original scales object applied after the additional and base config are combined to maintain original config.
This is a temporary solution as in v3 of chartjs 'scales' structure is changed into an object [rather than an Array] allowing me to clean this code up after the upgrade.

## Test instructions

Please go to Dashboard > 'Reports' > 'Diabetes'

Check localhost against production for visual discrepancies between all of the follow charts:
- Blood sugar <200 (line graph)
- Registrations and follow-ups (line graph and bar combo)
- Blood sugar 200-299 or ≥300 (stack bar chat with axes)
- Missed visits (line graph)
- Lost to follow-up (line graph) - See detail below for this graph

Lost to follow-up (line graph) is a single implementation for both HTN and DM.
Please go to Dashboard > 'Reports' > 'Hypertension' and check that graph renders as expected
